### PR TITLE
Code review Advanced-Building-Blocks-

### PR DESCRIPTION
**Pull request by JuanPabloGil**

### CODE REVIEW

File: methods.rb [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  def my_inject(initial = 0)
+    i = 0
+    accumulator = initial
+    while i < self.length
+      accumulator = yield(accumulator, self[i])
+      i += 1
+    end
+    accumulator
+  end
```
Comment (Ispirett) [inline methods.rb]: These methods are giving an error ` Failure/Error: yield i
     
     LocalJumpError:
       no block given (yield)
`

File: methods.rb [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  def my_any?
+    self.my_each { |i| return true if yield(i) }
+    false
+  end
+
+  def my_none?
+    self.my_each { |i| return false if yield(i) }
+    true
+  end
+
```
Comment (Ispirett) [inline methods.rb]: these methods also give this   
  ```ArgumentError:
       wrong number of arguments (given 1, expected 0) 
```
You need to give your method's parameters

File: methods.rb [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+arrayx = %w[jasa Joh Jean Jelean]
+puts arrayx.my_count
+puts 'my_map'
+arrayx = %w[jasa Joh Jean Jelean].my_map { |x| x + '!' }
+puts arrayx
+puts 'my_inject'
+def multiply_els(array)
+  array.my_inject(1) { |total, x| total * x }
+end
+puts multiply_els([2, 2, 4])
```
Comment (Ispirett) [inline methods.rb]: You  can remove this  code

File: methods.rb [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    else
+      for i in self
+        element = yield i
+        arry << element
+      end
+    end
+    arry
+  end
+
+  def my_inject(initial = 0)
```
Comment (Ispirett) [inline methods.rb]: this method also throws this errror    
```NoMethodError:
        undefined method `length' for 5..50:Range
```

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey @JuanPabloGil really great job :100: 

There are just some bugs  to fix :bug: 

Remember to write descriptive commits and pull request messages

File: methods.rb (hunk 4 above)
Comment (JuanPabloGil) [reply methods.rb]: Tested with `(5..10).inject(1) { |product, n| product * n }` & `(5..10).inject { |sum, n| sum + n } ` and works now

File: methods.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
+module Enumerable
+  def my_each
```
Comment (mcrd25) [inline methods.rb:5]: Additional Requirements for this method:
it should:
- return an Enumerator if no block is given

File: methods.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    for i in self
+      yield(i) if block_given?
+    end
+  end
+
+  def my_each_with_index
```
Comment (mcrd25) [inline methods.rb:16]: Additional Requirements for this method:
it should:
- return an Enumerator if no block is given

File: methods.rb [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    i = 0
+    for j in self
+      yield(j, i) if block_given?
+      i += 1
+    end
+  end
+
+  def my_select
```
Comment (mcrd25) [inline methods.rb:24]: Additional Requirements for this method:
it should:
- return an Enumerator if no block is given

File: methods.rb [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    result = []
+    self.my_each do |i|
+      result.<<(i) if yield(i)
+    end
+    result
+  end
+
+  def my_all?
```
Comment (mcrd25) [inline methods.rb]: Additional Requirements for this method:
it should:
- return true when no block or argument is given and none of the collection members are false or nil
- return true when a class is passed as an argument and all of the collection is a member of such class
- return true when a Regex is passed as an argument and all of the collection matches the Regex
- return true when a pattern other than Regex or a Class is given and if all of the collection matches the pattern

File: methods.rb [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    return true if !block_given?
+    if self.is_a? Hash
+      self.my_each { |m, n| return false if !yield(m, n) } 
+    else
+      self.my_each { |i| return false if !yield(i) }
+    end
+    true
+  end
+
+  def my_any?
```
Comment (mcrd25) [inline methods.rb]: Additional Requirements for this method:
it should:
- return true when no block or argument is given and at least one of the collection members are false or nil
- return true when a class is passed as an argument and at least one of the collection is a member of such class
- return true when a Regex is passed as an argument and at least one of the collection matches the Regex
- return true when a pattern other than Regex or a Class is given and if at least one of the collection matches the pattern

File: methods.rb [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    return true if !block_given?
+    if self.is_a? Hash
+      self.my_each { |m, n| return true if yield(m, n) }
+    else
+      self.my_each { |i| return true if yield(i) }
+    end
+    false
+  end
+
+  def my_none?
```
Comment (mcrd25) [inline methods.rb]: Additional Requirements for this method:
it should:
- return true when no block or argument is given and none of the collection members are true
- return true when a class is passed as an argument and not one of the collection is a member of such class
- return true when a Regex is passed as an argument and not one of the collection matches the Regex
- return true when a pattern other than Regex or a Class is given and if not one of the collection matches the pattern

File: methods.rb [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+      end
+    else
+      self.my_each do |i|
+        return false if yield(i)
+      end
+    end
+    true
+  end
+
+  def my_count?
```
Comment (mcrd25) [inline methods.rb]: This method does not return a boolean. It should be named `my_count`. It's not asking a question it's simply giving an answer.

Additional Requirements for this method:
it should:

- return the count of the number of items in enum that are equal to item if an argument is given

File: methods.rb [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    return self.length unless block_given?
+    count = 0
+    self.my_each { |i| count += 1 if yield(i) }
+    count
+  end
+
+  def my_map(proc = nil)
```
Comment (mcrd25) [inline methods.rb]: Additional Requirements for this method:
it should:
- return an Enumerator if no block is given

File: methods.rb [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    else
+      for i in self
+        element = yield(i)
+        arry << element
+      end
+    end
+    arry
+  end
+
+  def my_inject(*start_num)
```
Comment (mcrd25) [inline methods.rb]: Your inject does not pass any of our tests 😢 

It should:
- return combination of all elements of enum by applying a binary operation, specified by a block
- return combination of each element of the collection by applying the symbol as a named method when a symbol is specified

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes required


Hi @JuanPabloGil 👋,

Great attempt so far! Really good attempt. 🙂 
I also commend that you started using more descriptive commit messages 👏 You may check [this](https://chris.beams.io/posts/git-commit/) great guide out when you got time.

You currently have 21 failing tests. The reason for this is that your enumerable methods do not cover 
 all the scenarios covered in the original.
Please check out [this link](https://ruby-doc.org/core-2.6.3/Enumerable.html) for each enumerable you must implement and have it behave like the originals. For example what happens when a block is not given? Etc.
https://ruby-doc.org/core-2.6.3/Enumerable.html

For instance, you'd see `all` method behaves in the following way:
- it returns true if block never returns false or nil (you got this covered)
- If the block is not given, Ruby adds an implicit block of `{ |obj| obj }` which will cause `all?` to return `true` when none of the collection members are `false` or `nil`.
- If instead a pattern is supplied, the method returns whether pattern === element for every collection member ( **Extra Tip:** Same goes for class and regular expression

So if you cover those mentioned you'd be fine with tests.

As it stands non of your methods actually work fully. Please address this and submit for rereview.

Also, maybe you can try and update your current PR:
- Try to put a brief summary description in your pull request of all the changes you've made.
For example, I have an app that I set up and I added user model, comment model, post models, and related associations and migrations. I would mention this in my pull request description
```markdown
# Summary of changes made
- Generated User, Comment and Post Models
- Created Associations between User, Comment and Post Models
- Whatever other change you'd have made
```
**See review comments for more specific changes requested**

**Another thing~**
Please use [Git Flow](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) in next project!
***How Git Flow Works***
  - You'd have the `master` branch purely for production purposes 
  - the `development` branch would act as the main working branch where you create merge requests from `feature` branches to `development` for reviews on these requests
  - If you are requesting a review for a project without multiple milestones simply have one main `feature` branch (appropriately named) that you can then branch off from and merge to accordingly as you complete your project. 
  - After you have completed your project and it's ready for your first review you can then request a code review form your main `feature` branch to `development` branch
  - Once the pull request is approved you can then merge this pull request
  - You would then merge the `development` branch to `master` branch 
<br>
On another note, congrats on having a good READ ME and having your linter pass all tests! 👍 
<br>

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required. But, we are getting there! 💪 

Hi @JuanPabloGil 👋,

You did a great job! Great improvement! 😄 

There are now only 6 failing tests.
This involves 3 similar problems in `my_all`,  `my_any` and `my_none`

You need to consider when Class and Regular Expression is passed for each of them. It can be done similar to how it was done with a pattern. You may even consider the use of `===` if you are at a loss. But consider finding a way to test for Class and Regular expression passed as an argument. See my previous comments on this matter.


Once you address this you should be fine and get the 👍 


Great job again! 👏 

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

File: methods.rb [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    all = true
+    if block_given?
+      my_each { |x| unless yield(x) then all = false; break end }
+    elsif param
+      my_each { |x| return false unless pattern(x, param) }
+    else
+      my_each { |x| unless x then all = false; break end }
+    end
+    all
+  end
```
Comment (maelfosso) [inline methods.rb]:
```ruby
irb(main):006:0> array
=> [3, 4, 7, 6, 0, 8, 6, 0, 1, 3, 2, 0, 0, 5, 6, 2, 0, 7, 7, 2, 2, 8, 6, 8, 3, 5, 5, 0, 6, 2, 5, 6, 2, 3, 1, 2, 8, 5, 5, 7, 7, 7, 7, 5, 7, 5, 8, 2, 7, 4, 5, 2, 7, 2, 3, 1, 6, 3, 8, 4, 8, 6, 8, 3, 0, 4, 8, 2, 1, 4, 4, 3, 2, 2, 6, 8, 6, 7, 5, 7, 4, 6, 1, 2, 5, 1, 5, 4, 4, 4, 2, 0, 4, 2, 7, 4, 4, 5, 1, 5]
irb(main):007:0> array.all?(Integer)
=> true
irb(main):008:0> array.my_all?(Integer)
Traceback (most recent call last):
        5: from /home/elvisfosso/.rbenv/versions/2.5.1/bin/irb:11:in `<main>'
        4: from (irb):8
        3: from methods.rb:37:in `my_all?'
        2: from methods.rb:11:in `my_each'
        1: from methods.rb:37:in `block in my_all?'
NoMethodError (undefined method `pattern' for #<Array:0x000055bafc8ffec8>)
Did you mean?  pattern?
irb(main):009:0>
```

I think you wanted to say **pattern?**. With that there is no error.
Please, fix it

File: methods.rb [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    any = false
+    if block_given?
+      my_each { |x| if yield(x) then any = true; break end }
+    elsif param
+      my_each {|x | return true if pattern(x, param) }
+    else
+      my_each { |x| if x then any = true; break end }
+    end
+    any
+  end
```
Comment (maelfosso) [inline methods.rb]: Do you mean `pattern` or **`pattern?`** ? 
Please, fix that

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @JuanPabloGil 

You did a really awesome job :smile: 
You're about to finish. Just only one bug to fix. Change `pattern` to **`pattern?`** in `my_all?` and `my_any?` methods.

Please, make that only change and submit a new code review request.
Thanks.

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap: You did it :smile: 

**The Project is approved** :+1: 
You can merge.

Good luck with the next project.

**Pull request merged**